lib/plugin/selenoid.js

```diff
@@ -280,11 +236,128 @@ const selenoid = (config) => {
     }
   });
 
-  if (deletePassed) {
+  if (config.enableVideo) {
     event.dispatcher.on(event.test.passed, (test) => {
-      passedTests.push(test.title);
+      if (deletePassed) {
+        passedTests.push(test.title);
+      } else {
+        videoSaved(test);
+      }
+    });
+
+    event.dispatcher.on(event.test.failed, (test) => {
+      videoSaved(test);
     });
   }
 };
 
 module.exports = selenoid;
+
+function videoSaved(test) {
+  const fileName = `${clearString(test.title)}.mp4`;
+  output.debug(`Video has been saved to file://${global.output_dir}/video/${fileName}`);
+  const allureReporter = container.plugins('allure');
+  if (allureReporter) {
+    allureReporter.addAttachment('Video', fs.readFileSync(path.join(global.output_dir, 'video', fileName)), 'video/mp4');
+  }
+}
+
+const createSelenoidConfig = () => {
+  const configPath = path.join(global.codecept_dir, 'browsers.json');
+  return new Promise((res, rej) => {
+    try {
+      if (fs.existsSync(configPath)) {
+        res(true);
+      } else {
+        const data = new Uint8Array(Buffer.from(JSON.stringify(defaultBrowserConfig)));
+        fs.writeFileSync(configPath, data);
+        res(true);
+      }
+    } catch (err) {
+      rej(err.stack);
+    }
+  });
+};
+
+const createSelenoid = () => exec(dockerCreateScript);
+
+const startSelenoid = () => exec(dockerStartScript);
+
+const stopSelenoid = () => exec(dockerStopScript);
+
+const checkDockerImage = () => exec(dockerImageCheckScript);
+
+const pullImage = async () => {
+  const { stdout } = await checkDockerImage();
+  const pulls = [];
+  let resultPromise;
+
+  output.print('Pulling in Selenoid containers. This may take a while...');
+
+  console.time('Pulled containers');
+  if (!stdout.includes('selenoid/video-recorder')) {
+    output.debug('Pulling selenoid/video-recorder...');
+    resultPromise = exec('docker pull selenoid/video-recorder:latest-release')
+      .then(() => output.debug('Pulled in selenoid/video-recorder'));
+    pulls.push(resultPromise);
+  }
+  if (!stdout.includes('selenoid/chrome')) {
+    output.debug('Pulling selenoid/chrome...');
+    resultPromise = exec('docker pull selenoid/chrome:latest')
+      .then(() => output.debug('Pulled in selenoid/video-recorder'));
+    pulls.push(resultPromise);
+  }
+  if (!stdout.includes('selenoid/firefox')) {
+    output.debug('Pulling selenoid/firefox...');
+    resultPromise = exec('docker pull selenoid/firefox:latest')
+      .then(() => output.debug('Pulled in selenoid/chrome'));
+    pulls.push(resultPromise);
+  }
+
+  return Promise.all(pulls).then(() => {
+    console.timeEnd('Pulled containers');
+    output.debug('Disable `autoCreate` in Selenoid plugin and pull in all containers manually to reduce this time');
```

Comment:
This will take time on first time alone since next time images will be already present and it won't pull the image again. :)

Reply:
I agree but even pulling and checking images always takes about 2 secs for me. 
Not the big deal but I think it is useful to state that this amount of time can be reduced. Anyway, it is printed in debug mode only...

Reply:
Anyway, I need to write this better...